README-PLUGINS.md

```diff
@@ -0,0 +1,92 @@
+# nbgitpuller - downloader plugin documentation
+
+nbgitpuller uses [pluggy](https://pluggy.readthedocs.io/en/stable/) as a framework
+to load any installed nbgitpuller-downloader plugins. There are three downloader plugins
+available right now:
+- [nbgitpuller-downloader-googledrive](https://github.com/jupyterhub/nbgitpuller-downloader-googledrive)
+- [nbgitpuller-downloader-dropbox](https://github.com/jupyterhub/nbgitpuller-downloader-dropbox)
+- [nbgitpuller-downloader-generic-web](https://github.com/jupyterhub/nbgitpuller-downloader-generic-web)
+
+
+There are several pieces to be aware of for the plugin to work correctly:
+1. The setup.cfg(or setup.py) file must have the entry_points definition.
+For example:  
+> [options.entry_points]  
+nbgitpuller =  
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;dropbox=nbgitpuller_downloader_dropbox.dropbox_downloader
+
+2. The file referenced for use by nbgitpuller in the plug-in (the above example is looking for the 
+file, dropbox_downloader) must implement the function handle_files(query_line_args) and be decorated with `@hookimpl`.
+3. As a consequence of this, the following must be imported:
+    - `from nbgitpuller.hookspecs import hookimpl`
+4. The implementation of the handle_files function in your plugin needs to return
+   two pieces of information:
+   - the name of the folder, the archive is in after decompression
+   - the path to the local git repo mimicking a remote origin repo
+   
+nbgitpuller provides a function in plugin_helper.py called handle_files_helper that handles the downloading
+and returning of the correct information if given a URL, the extension of the
+file to decompress(zip or tar) and the progress function(I will describe that
+more later) but you are welcome to implement the functionality of handle_files_helper in your
+plug-in. There may be use cases not covered by the currently available plugins like needing to authenticate against
+the webserver or service where your archive is kept. Either way, it behooves you
+to study the handle_files_helper function in nbgitpuller to get a sense of how this function
+is implemented.
+
+For the rest of the steps, I refer you to the [nbgitpuller-downloader-dropbox](https://github.com/jupyterhub/nbgitpuller-downloader-dropbox) plugin.  
+   ```
+   @hookimpl  
+   def handle_files(query_line_args):
+      query_line_args["repo"] = query_line_args["repo"].replace("dl=0", "dl=1")  # dropbox: dl set to 1  
+      ext = determine_file_extension(query_line_args["repo"])`  
+      query_line_args["extension"] = ext
+      loop = asyncio.get_event_loop()
+      tasks = handle_files_helper(query_line_args), query_line_args["progress_func"]()
+      result_handle, _ = loop.run_until_complete(asyncio.gather(*tasks))
+      return result_handle
+   ```
```

Comment:
Fixed indentation in the python code and in the markdown file, and made it syntax highlight as python.

````suggestion

```python
@hookimpl  
def handle_files(query_line_args):
    query_line_args["repo"] = query_line_args["repo"].replace("dl=0", "dl=1")  # dropbox: dl set to 1  
    ext = determine_file_extension(query_line_args["repo"])`  
    query_line_args["extension"] = ext
    loop = asyncio.get_event_loop()
    tasks = handle_files_helper(query_line_args), query_line_args["progress_func"]()
    result_handle, _ = loop.run_until_complete(asyncio.gather(*tasks))
    return result_handle
```
````